### PR TITLE
RISCV: Add some definitions and vararg supports

### PR DESCRIPTION
**Pull request by Ast-x64**
This is based on RISCV Calling Convention and hints in comments (e.g.
	elf/elf.h in glibc).

### CODE REVIEW

**Review by ljmf00 — APPROVED** [review]
Looks good. Anyone with experience in the RISC-V ABI to double-check this?

File: src/rt/dwarfeh.d [hunk 1]
```diff
@@ -73,6 +73,11 @@ else version (MIPS32)
     enum eh_exception_regno = 4;
     enum eh_selector_regno = 5;
 }
+else version (RISCV64)
```
Comment (ljmf00) [inline src/rt/dwarfeh.d:76]: Perhaps, shouldn't be available a `regno` for `RISCV32` version?

Comment (ljmf00) [reply src/rt/dwarfeh.d:76]: It seems to be the same, according to this LLVM diff https://reviews.llvm.org/D63411

Comment (Ast-x64) [reply src/rt/dwarfeh.d:76]: Thanks for your review. I'm currently working on riscv64gc native environment, and I don't have a native riscv32 gcc available now to execute the test code in the comment above. Maybe someone more experienced could help?

Comment (ljmf00) [reply src/rt/dwarfeh.d:76]: It seems to be correct, according to both LLVM and RISC-V ABI docs. See https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-dwarf.adoc#dwarf-register-numbers . `X10` and `X11` maps directly to DWARF registers, so it should be good for both architectures according to the LLVM patch. Also, according to the comments on that diff, they based those values on GCC, so it should match, but the RISC-V ABI documentation should mention it explicitly. I can file an issue on it to clarify those values.

Comment (ljmf00) [reply src/rt/dwarfeh.d:76]: Cross referencing: https://github.com/riscv-non-isa/riscv-elf-psabi-doc/issues/293

File: src/core/sys/elf/package.d [hunk 2]
```diff
@@ -2510,3 +2510,66 @@ enum R_TILEGX_GNU_VTINHERIT =  128;
 enum R_TILEGX_GNU_VTENTRY =    129;
 
 enum R_TILEGX_NUM =            130;
+
+enum EF_RISCV_RVC =            0x0001;
+enum EF_RISCV_FLOAT_ABI =      0x0006;
+enum EF_RISCV_FLOAT_ABI_SOFT = 0x0000;
+enum EF_RISCV_FLOAT_ABI_SINGLE = 0x0002;
```
[the remark this answers is not in the document]
Comment (Ast-x64) [reply src/core/sys/elf/package.d:2517]: Thanks for your review. I've just re-aligned these columns in https://github.com/dlang/druntime/pull/3846/commits/4b9f519b2c7fd62f4fc8efa89f5ad014f448be5d.

**Comment by ibuclaw** [comment]
> Looks good. Anyone with experience in the RISC-V ABI to double-check this?

va_list is indeed a `void*` on RISC-V.  Can't say much about the other parts of varargs and dwarfeh, other than seem reasonable. I assume this is for the benefit of LDC?  (GDC has `__builtin_eh_return_data_regno` and `__builtin_va_arg` intrinsics exposed from the backend)

**Comment by Ast-x64** [comment]
> I assume this is for the benefit of LDC?

Yes I'm trying to port LDC to riscv64gc platform. My current draft version shows, `printf` (which uses vararg to pass parameters) is working properly under my tests.

**Pull request merged**